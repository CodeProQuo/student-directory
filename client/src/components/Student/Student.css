/* .student styling */

.student {
    border-bottom: 1px solid lightgrey;
    padding-bottom: 10px;
    display: flex;
    flex-direction: row;
}

.student h1 {
    margin: 10px 0 15px 0;
    font-family: Raleway bold, sans-serif;
}

.student img {
    border: 1px solid lightgrey;
    border-radius: 50%;
    width: 100px;
    margin: 20px;
}

.collapsed {
    display: none;
}

.student-info {
    display: flex;
    flex-direction: column;
    flex-grow: 1;
}

/* .student-details styling */

.student-details {
    margin-top: 10px;
}

/* tag styling */

.tag-container {
    display: flex;
    flex-wrap: wrap;
    margin: 5px 0 5px 20px;
}

.tag {
    background-color: #dcdcdc;
    border-radius: 5px;
    display: inline;

    margin: 2px 2px 2px 2px;
    padding: 5px 5px 5px 5px;
}

.tag:hover {
    background-color: gray;
}

/* .expand-btn styling*/

.expand-btn {
    background-color: white;
    border: 0;
    height: 0;
}

.expand-btn:focus {
    outline: 0;
}

.expand-btn::after {
    content: "\2212";
    color: grey;
    font-weight: bold;
    font-size: 40px;
}

.expand-btn.closed::after {
    content: "\002B";
}

.expand-btn:hover:after {
    color: black;
}

/* #add-tag-input styling */

#add-tag-input {
    width: 100%;
    border: 0;
    border-bottom: 1px solid #dcdcdc;
    height: 30px;
    padding: 10px 0 0 0;
    font-size: 15px;
}

#add-tag-input:focus {
    outline: none;
    border-bottom: 1px solid black;
}

