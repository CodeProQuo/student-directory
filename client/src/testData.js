export const firstStudent = {
    pic: "http://www.google.com",
    firstName: "John",
    lastName: "Doe",
    email: "johndoe@email.com",
    company: "ABC",
    skill: "Archery",
    grades: ["20", "30", "40"],
};

export const secondStudent = {
    pic: "http://www.google.com",
    firstName: "Jane",
    lastName: "Doe",
    email: "janedoe@email.com",
    company: "NBC",
    skill: "Bowling",
    grades: ["50", "60", "70", "80"],
};