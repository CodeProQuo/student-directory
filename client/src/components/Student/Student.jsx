import React, {useState, useEffect} from 'react';
import "./Student.css";

const Student = (props) => {

    const [expanded, setExpanded] = useState(false);
    const [tags, setTags] = useState([]);
    const [passesFilter, setPassesFilter] = useState(true);

    const {nameFilter, tagFilter, student} = props;

    const handleClickExpand = () => {
        setExpanded(!expanded);
    };

    const handleKeyDown = (e) => {
        if (e.key === 'Enter' && e.target.value !== "") {
            setTags([...tags, e.target.value]);
            e.target.value = "";
        }
    };

    const handleTagRemove = (e) => {
        setTags(tags.filter((tag) => tag !== e.target.textContent));
    };

    const {grades, pic, firstName, lastName, email, company, skill} = student;
    const average = grades.reduce((gradesSum, grade) => gradesSum + parseInt(grade), 0) / grades.length;

    const testGrades = grades.map((grade, index) => {
        return <p key={index} >Test {index + 1}: {grade}%</p>
    });

    const tagJsx = tags.map((tag, index) => {
        return <div key={index} className="tag" onClick={handleTagRemove}>{tag}</div>
    });

    useEffect(()=> {
        const filterCheck = `${firstName} ${lastName}`.toLowerCase().includes(nameFilter.toLowerCase()) &&
            (tagFilter === "" || tags.filter((tag) => tag.toLowerCase().includes(tagFilter.toLowerCase())).length);

        setPassesFilter(filterCheck);
    }, [nameFilter, tagFilter, firstName, lastName, tags]);

    return (
        <div data-testid="student" className={passesFilter ? "student" : "student collapsed"}>
            <div>
                <img src={pic} alt="Student's picture"/>
            </div>
            <div className="student-info">
                <h1>{firstName.toUpperCase()} {lastName.toUpperCase()}</h1>
                <p>Email: {email}</p>
                <p>Company: {company}</p>
                <p>Skill: {skill}</p>
                <p>Average: {average}%</p>
                <div className={expanded ? "student-details" : "student-details collapsed"}>
                    {testGrades}
                    <div className="tag-container">
                        {tagJsx}
                    </div>
                    <input type="text" onKeyDown={handleKeyDown} id="add-tag-input" placeholder="Add new tag"/>
                </div>
            </div>
            <button onClick={handleClickExpand}
                    className={expanded ? "expand-btn" : "expand-btn closed"}/>
        </div>);
};

export default Student;
