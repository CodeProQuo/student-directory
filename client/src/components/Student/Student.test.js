import React from 'react';
import fs from 'fs';
import path from 'path';
import {render, cleanup, fireEvent} from '@testing-library/react';
import Student from './Student';
import {firstStudent, secondStudent} from "../../testData";

const cssFile = fs.readFileSync(
    path.resolve(__dirname, './Student.css'),
    'utf8'
);

const style = document.createElement('style');
style.type = 'text/css';
style.innerHTML = cssFile;

afterEach(cleanup);

it('should take a snapshot of Student', () => {
    const {asFragment} = render(<Student nameFilter=""
                                         tagFilter=""
                                         student={firstStudent}/>);

    expect(asFragment(<Student/>)).toMatchSnapshot();
});

it('should initialize states for Student as intended', () => {
    const {getByTestId, getByPlaceholderText, container} = render(<Student nameFilter=""
                                                                           tagFilter=""
                                                                           student={firstStudent}/>);

    container.append(style);

    expect(getByTestId("student")).toBeVisible();
    expect(getByPlaceholderText('Add new tag')).not.toBeVisible();
    expect(getByTestId("student")).not.toContainElement(container.querySelector("[class='tag']"));
});

it('should render correctly with all its props', () => {
    const {container} = render(<Student nameFilter=""
                                        tagFilter=""
                                        student={firstStudent}/>);

    expect(container).toHaveTextContent("JOHN DOE");
    expect(container).toHaveTextContent("Email: johndoe@email.com");
    expect(container).toHaveTextContent("Company: ABC");
    expect(container).toHaveTextContent("Skill: Archery");

    expect(container).toHaveTextContent("Test 1: 20%");
    expect(container).toHaveTextContent("Test 2: 30%");
    expect(container).toHaveTextContent("Test 3: 40%");
});

it('should show student details on clicking expand button', async () => {
    const {container, getByPlaceholderText, getByRole, getByText} = render(<Student nameFilter=""
                                                                                    tagFilter=""
                                                                                    student={firstStudent}/>);

    container.append(style);

    expect(getByPlaceholderText('Add new tag')).not.toBeVisible();
    expect(getByText("Test 1: 20%")).not.toBeVisible();

    fireEvent.click(getByRole('button'));

    expect(getByPlaceholderText('Add new tag')).toBeVisible();
    expect(getByText("Test 1: 20%")).toBeVisible();
});


it('should add tag', () => {
    const {getByPlaceholderText, getByRole, getByText, container} = render(<Student nameFilter=""
                                                                                    tagFilter=""
                                                                                    student={firstStudent}/>);
    container.append(style);

    fireEvent.click(getByRole('button'));

    const addTagInput = getByPlaceholderText('Add new tag');
    fireEvent.change(addTagInput, {target: {value: 'Tag1'}});
    fireEvent.keyDown(addTagInput, {key: 'Enter', keyCode: 13});

    expect(getByText("Tag1")).toBeVisible();
});

it('should remove tag', () => {
    const {
        getByPlaceholderText,
        getByRole,
        getByText,
        queryByText,
        container
    } = render(<Student nameFilter="" tagFilter="" student={firstStudent}/>);

    container.append(style);

    fireEvent.click(getByRole('button'));

    const addTagInput = getByPlaceholderText('Add new tag');
    fireEvent.change(addTagInput, {target: {value: 'Tag1'}});
    fireEvent.keyDown(addTagInput, {key: 'Enter', keyCode: 13});
    fireEvent.click(getByText("Tag1"));

    expect(queryByText("Tag1")).not.toBeInTheDocument();

});

it('should calculate the grade average correctly', () => {
    const {container} = render(<Student nameFilter=""
                                        tagFilter=""
                                        student={secondStudent}/>);

    expect(container).toHaveTextContent("Average: 65%");
});

it('should filter by name', () => {
    const nameFilter = "Jane";
    const {getByText, container} = render(<Student nameFilter={nameFilter}
                                                   tagFilter=""
                                                   student={firstStudent}/>);
    render(<Student nameFilter={nameFilter}
                    tagFilter=""
                    student={secondStudent}/>);

    container.append(style);

    expect(getByText("JANE DOE")).toBeVisible();
    expect(getByText("JOHN DOE")).not.toBeVisible();
});

it('should filter by tag', () => {
    const {
        getByText,
        container,
        getAllByRole,
        getAllByPlaceholderText,
        rerender
    } = render(<Student nameFilter="" tagFilter="" student={firstStudent}/>);

    render(<Student nameFilter=""
                    tagFilter=""
                    student={secondStudent}/>);

    container.append(style);

    const expandButton = getAllByRole('button');

    fireEvent.click(expandButton[0]);

    const addTagInputAll = getAllByPlaceholderText('Add new tag');

    fireEvent.change(addTagInputAll[0], {target: {value: 'Tag1'}});
    fireEvent.keyDown(addTagInputAll[0], {key: 'Enter', keyCode: 13});


    fireEvent.click(expandButton[1]);
    fireEvent.change(addTagInputAll[1], {target: {value: 'Tag2'}});
    fireEvent.keyDown(addTagInputAll[1], {key: 'Enter', keyCode: 13});

    rerender(<Student nameFilter=""
                      tagFilter="Tag2"
                      student={firstStudent}/>);
    rerender(<Student nameFilter=""
                      tagFilter="Tag2"
                      student={secondStudent}/>);

    expect(getByText("Tag2")).toBeVisible();
    expect(getByText("Tag1")).not.toBeVisible();
});

it('should filter by name and tag', () => {
    const {
        getAllByTestId,
        container,
        getAllByRole,
        getAllByPlaceholderText,
        rerender: rerenderFirst
    } = render(<Student nameFilter="" tagFilter="" student={firstStudent}/>);

    const {rerender: rerenderSecond} = render(<Student nameFilter=""
                                                       tagFilter=""
                                                       student={secondStudent}/>);

    const {rerender: rerenderThird} = render(<Student nameFilter=""
                                                      tagFilter=""
                                                      student={firstStudent}/>);

    container.append(style);

    const expandButton = getAllByRole('button');

    fireEvent.click(expandButton[0]);

    const addTagInputAll = getAllByPlaceholderText('Add new tag');

    fireEvent.change(addTagInputAll[0], {target: {value: 'Tag1'}});
    fireEvent.keyDown(addTagInputAll[0], {key: 'Enter', keyCode: 13});


    fireEvent.click(expandButton[1]);
    fireEvent.change(addTagInputAll[1], {target: {value: 'Tag1'}});
    fireEvent.keyDown(addTagInputAll[1], {key: 'Enter', keyCode: 13});

    rerenderFirst(<Student nameFilter="John"
                           tagFilter="Tag1"
                           student={firstStudent}/>);
    rerenderSecond(<Student nameFilter="John"
                            tagFilter="Tag1"
                            student={secondStudent}/>);
    rerenderThird(<Student nameFilter="John"
                           tagFilter="Tag1"
                           student={firstStudent}/>);

    const studentComponents = getAllByTestId("student");

    expect(studentComponents[0]).not.toHaveClass("collapsed");
    expect(studentComponents[1]).toHaveClass("collapsed");
    expect(studentComponents[2]).toHaveClass("collapsed");
});