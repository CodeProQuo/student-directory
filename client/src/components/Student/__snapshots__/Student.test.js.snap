// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot of Student 1`] = `
<DocumentFragment>
  <div
    class="student"
    data-testid="student"
  >
    <div>
      <img
        alt="Student's picture"
        src="http://www.google.com"
      />
    </div>
    <div
      class="student-info"
    >
      <h1>
        JOHN DOE
      </h1>
      <p>
        Email: johndoe@email.com
      </p>
      <p>
        Company: ABC
      </p>
      <p>
        Skill: Archery
      </p>
      <p>
        Average: 30%
      </p>
      <div
        class="student-details collapsed"
      >
        <p>
          Test 1: 20%
        </p>
        <p>
          Test 2: 30%
        </p>
        <p>
          Test 3: 40%
        </p>
        <div
          class="tag-container"
        />
        <input
          id="add-tag-input"
          placeholder="Add new tag"
          type="text"
        />
      </div>
    </div>
    <button
      class="expand-btn closed"
    />
  </div>
</DocumentFragment>
`;
